src/clojure/zero_one/geni/arrow.clj

```diff
@@ -0,0 +1,222 @@
+(ns zero-one.geni.arrow
+  (:import [org.apache.arrow.vector VarCharVector
+            VectorSchemaRoot BigIntVector Float4Vector Float8Vector IntVector
+            BitVector ValueVector BaseFixedWidthVector TimeStampMilliVector]
+           [org.apache.arrow.vector.util Text]
+           [org.apache.arrow.vector.ipc ArrowStreamWriter]
+           [org.apache.arrow.memory RootAllocator]
+           [java.nio.channels Channels]
+           [org.apache.spark.sql Row]
+           [org.apache.arrow.vector.types.pojo Schema]
+           [scala.collection.convert Wrappers$IteratorWrapper]
+           )
+  (:require
+   [clojure.java.io :as io]
+
+   ))
+
+;; (set! *warn-on-reflection* true)
+
+
+;;;  these protocolls are just needed to make type hints working for the typed-action function
+(defprotocol IValueInfo
+  (^BaseFixedWidthVector vector [this])
+  (^long idx1 [this])
+  (^Object value [this])
+  )
+
+(defprotocol IRowInfo
+  (^Row row [this])
+  (^long idx2 [this])
+  )
+
+(defrecord ValueInfo [^BaseFixedWidthVector a-v ^long a-idx ^Object a-value]
+  IValueInfo
+  (vector ^BaseFixedWidthVector [this] a-v)
+  (idx1 ^long [this] a-idx)
+  (value ^Object [this] a-value)
+
+  )
+
+
+
+(defrecord RowInfo [^Row row ^long idx]
+  IRowInfo
+  (row ^Row [this] row )
+  (idx2 ^long [this] idx)
+  )
+
+;;; this method will be called for evey value transformed, so relection shoudkl be avoided
+(defn typed-action [action type value-info row-info ^String name ^RootAllocator allocator]
+  (when (= :set type)
+    (when (nil? (value value-info))
+      (throw (Exception. "value cannot be null"))
+      ))
+  (case type
```

Comment:
I think we I would prefer a nested map of functions, instead of the nested cases, but yea, this is a matter of taste I suppose. I'm happy with this if you feel strongly about it :)

Reply:
Does a "map of functions" work with  "calling java functions" ?
I remembered that we cannot do something like

(def m (hash-map f (.set))

The whole thing became as well "ugly", because I needed to have type-hints working.
They makes a huge difference in performance.


If you have an other proposal, how I can re-write it using:
- a map of functions
- type hints

On the other side, I am quite happy, that "missing" types now result in IllegalArgumentException.

Reply:
code is far better now.